util/attendance/src/lib.rs

```diff
@@ -0,0 +1,116 @@
+#![no_std]
+
+#[allow(unused_imports)]
+use multiversx_sc::imports::*;
+
+#[multiversx_sc::contract]
+pub trait Attendance: multiversx_sc_modules::only_admin::OnlyAdminModule {
+    #[init]
+    fn init(&self, admins: MultiValueEncoded<ManagedAddress>) {
+        self.admins().extend(admins);
+    }
+
+    #[only_owner]
+    #[upgrade]
+    fn upgrade(&self) {}
+
+    #[endpoint(registerStudent)]
+    fn register_student(&self) {
+        let caller = self.blockchain().get_caller();
+
+        require!(
+            self.students().contains(&caller),
+            "Student is already registered"
+        );
+
+        self.students().insert(caller);
+    }
+
+    #[endpoint(registerAttendance)]
+    fn register_attendance(&self, secret_key: ManagedBuffer) {
+        let caller = self.blockchain().get_caller();
+        self.require_caller_registered(&caller);
+        self.check_secret_key(&secret_key, &caller);
+
+        self.attendance(&caller)
+            .update(|attendance| *attendance += 1);
+        self.cooldown(&caller)
+            .set(self.blockchain().get_block_epoch());
+    }
+
+    #[only_admin]
+    #[endpoint(registerSecretKey)]
+    fn register_secret_key(&self, secret_key: ManagedBuffer) {
+        let current_epoch = self.blockchain().get_block_epoch();
+        let secret_key_mapper = self.secret_key(&secret_key);
+
+        require!(
+            secret_key_mapper.is_empty(),
+            "This secret key was already registered"
+        );
+
+        secret_key_mapper.set(current_epoch);
+    }
+
+    #[only_admin]
+    #[endpoint(insertAdmin)]
+    fn insert_admin(&self, admin: ManagedAddress) {
+        self.require_admin_not_registered(&admin);
+        self.add_admin(admin);
+    }
+
+    fn check_secret_key(&self, secret_key: &ManagedBuffer, student: &ManagedAddress) {
+        let current_epoch = self.blockchain().get_block_epoch();
+
+        self.require_is_secret_correct(secret_key);
+        self.require_cooldown_has_passed(student, current_epoch);
+    }
+
+    fn require_cooldown_has_passed(&self, student: &ManagedAddress, current_epoch: u64) {
+        let student_cooldown = self.cooldown(student).get();
+
+        require!(
+            student_cooldown + 7 < current_epoch,
```

Comment:
7 epochs, means that if Thursday student calls `register_attendance`, he will be able to call it again on Friday. Maybe switch to 6.